src/roaring.c

```diff
@@ -87,45 +87,39 @@ bool roaring_bitmap_init_with_capacity(roaring_bitmap_t *r, uint32_t cap) {
-    uint32_t prev = 0;       // previous valued inserted
-    size_t i = 0;            // index of value
-    int containerindex = 0;
-    if (n_args == 0) return;
+    size_t i;
     uint32_t val;
-    memcpy(&val, vals + i, sizeof(val));
-    container =
-        containerptr_roaring_bitmap_add(r, val, &typecode, &containerindex);
-    prev = val;
-    i++;
-    for (; i < n_args; i++) {
-        memcpy(&val, vals + i, sizeof(val));
-        if (((prev ^ val) >> 16) ==
-            0) {  // no need to seek the container, it is at hand
-            // because we already have the container at hand, we can do the
-            // insertion
-            // automatically, bypassing the roaring_bitmap_add call
-            uint8_t newtypecode = typecode;
-            container_t *container2 =
-                container_add(container, val & 0xFFFF, typecode, &newtypecode);
-            if (container2 != container) {  // rare instance when we need to
-                                            // change the container type
-                container_free(container, typecode);
-                ra_set_container_at_index(&r->high_low_container,
-                                          containerindex, container2,
-                                          newtypecode);
-                typecode = newtypecode;
-                container = container2;
-            }
-        } else {
-            container = containerptr_roaring_bitmap_add(r, val, &typecode,
-                                                        &containerindex);
+    roaring_bulk_context_t context = {0};
+    for (i = 0; i < n_args; i++) {
+        memcpy(&val, &vals[i], sizeof(val));
+        roaring_bitmap_add_bulk(r, &context, val);
+    }
+}
+
+void roaring_bitmap_add_bulk(roaring_bitmap_t *r,
+                             roaring_bulk_context_t *context, uint32_t val) {
+    uint16_t key = val >> 16;
+    if ((context->key != key) || context->container == NULL) {
+        context->container = containerptr_roaring_bitmap_add(
+            r, val, &context->typecode, &context->idx);
+        context->key = val >> 16;
+    } else {
+        // no need to seek the container, it is at hand
+        // because we already have the container at hand, we can do the
+        // insertion directly, bypassing the roaring_bitmap_add call
+        uint8_t new_typecode = context->typecode;
+        container_t *container2 = container_add(
+            context->container, val & 0xFFFF, context->typecode, &new_typecode);
+        if (container2 != context->container) {
+            // rare instance when we need to change the container type
+            container_free(context->container, context->typecode);
+            ra_set_container_at_index(&r->high_low_container, context->idx,
+                                      container2, new_typecode);
+            context->typecode = new_typecode;
+            context->container = container2;
```

Comment:
Using the containers you typically use for adding elements one by one is unlikely to give as much gain as, say, allocating a bitmap container at first and only converting at flush time (when the key changes). Otherwise you may reallocate and move data a lot of times.
It would have a constant memory overhead tho, so benchmark first if you decide to try it.

Reply:
Because there's currently no destruction process for a `roaring_bitmap_context_t`, it would also possibly leave extra bitmap containers whenever you stop using a context, i.e. if you wrote:

```c
for (i = 0; i < 1000; i++) {
  roaring_bulk_context_t context = {0};
  roaring_bitmap_add_bulk(bm, &context, i * 0x10000);
}
```
That would leave behind 1000 bitmap containers, all with 1 bit set.

I'm still seeing a pretty nice win over bare `roaring_bitmap_add` as is. There is some extra padding still in `roaring_bulk_context_t`, we could choose to opt-in somehow and require a function be called after bulk operations to finish a context if it has a flag set, but I think that would a future effort.